### ISSUE TITLE
Uri and query keys

### ISSUE THREAD

**Issue by bryce-anderson** [http4s/http4s#129]
The new typeclass based query system is pretty nice for dealing with values, but the keying is not as nice IMO.

Currently there are three ways to add a query param starting [here](https://github.com/http4s/http4s/blob/master/core/src/main/scala/org/http4s/Uri.scala#L115). 

The keying based on type is nice for situations where you have complex objects you want to serialize a single object of that type (a custom case class ```Foo(bar: Int, n: String)``` for example) but simple key value stuff is a little second class. For example, rho also has a query param concept [here](https://github.com/http4s/rho/blob/master/examples/src/main/scala/com/http4s/rho/hal/plus/swagger/demo/RestService.scala#L20) that is not compatible with the current ```Uri``` query situation. This could be fixed by adding a method such as either
```scala
  def +?[T: QueryParamEncoder](name: QueryParam[T], values: T*): Uri =
    _withQueryParam(name.key, values map QueryParamEncoder[T].encode)

// or in the typeclass pattern. I'm sure this doesn't compile but the concept should work
def +?[T: QueryParamEncoder, K: QueryParamLike[K, T]](key: K, values: T*): Uri =
    _withQueryParam(QueryParamLike[K,T].key(key), values map QueryParamEncoder[T].encode)
```

The second one is probably the preferred but more complex solution, but either would allow me to use my query params in rho as first class keys.

**Comment by before** [comment]
Would it be an option to remove that [one](https://github.com/http4s/http4s/blob/topic/QueryParamKey/core/src/main/scala/org/http4s/Uri.scala#L118-L120) and to provide some documentation how to use the typeclass style key syntax prominently?

**Comment by julien-truffaut** [comment]
sorry I lost track of this, I cannot remember how `QueryParamKeyLike` is intended to be use? is it intended to merge both methods with a `String` key and a `QueryParameterKey`?

One way around it could be to remove the varargs completely and implement `QueryParamEncoder` for different data structure:

```scala
def +?[T: QueryParamEncoder, K: QueryParamKeyLike](key: K, value: T): Uri = ???

implicit def  listEncoder[T: QueryParamEncoder]: QueryParamEncoder[List[T]] = ???

uri +?("names", List("John", "Paul")) 
```
Syntax will be slightly worse but the number of methods will shrink a lot

**Comment by bryce-anderson** [comment]
I've let this one fall by the wayside a bit myself, I've been preoccupied getting HTTP/2.0 support into blaze.

@julien-truffaut, wes, the ```QueryParamKeyLike``` could be used to merge the ```String``` and ```QueryParameterKey``` receiving methods. If we look close, the only difference between the ```QueryParamKey``` and ```QueryParam``` types is that ```QueryParamKey``` has the ability to accept a value. Would it be a problem to just add the ability to receive a value to ```QueryParam``` and drop the ```QueryParamKey```? For cases where the type defines the key it can ignore the value and return a constant key, and for others it can extract a key from the received value.

The removal of the varargs might be a good idea to resolve the method resolution issue, certainly worth trying. I'll spend a little time working on my branch and see what happens.

**Comment by bryce-anderson** [comment]
Well, a quick look at the ```QueryParamKeyLike``` and ```QueryParam``` issue shows it doesn't work: methods like
```scala
def -?[T: QueryParam]: Uri = ???
```
wouldn't work because it has no val to feed to the hypothetical type signature.

**Comment by bryce-anderson** [comment]
Using ```QueryParamKeyLike``` to handle Strings is a nogo: I get collisions between 
```
def +?[T: QueryParam : QueryParamEncoder](values: T): Uri = ??? // T can be Seq[T2]
// and
def +?[K: QueryParamKeyLike](name: K): Uri = ???
```
No surprise.

**Comment by julien-truffaut** [comment]
we could also remove the last method as it is equivalent to call:

```scala
uri +?("key", Unit)
// or
uri +?("key", Nil) 
```

**Comment by bryce-anderson** [comment]
```uri +?("key", ()) ``` works nicely.

**Comment by bryce-anderson** [comment]
Closed via #161
